filebeat/input/filestream/prospector.go

```diff
@@ -115,23 +160,55 @@ func (p *fileProspector) Run(ctx input.Context, s *statestore.Store, hg loginp.H
 					}
 				}
 
-				hg.Run(ctx, src)
+				hg.Start(ctx, src)
 
 			case loginp.OpDelete:
 				log.Debugf("File %s has been removed", fe.OldPath)
 
+				if p.stateChangeCloser.Removed {
+					log.Debugf("Stopping harvester as file %s has been removed and close.on_state_change.removed is enabled.", src.Name())
+					hg.Stop(src)
+				}
+
 				if p.cleanRemoved {
 					log.Debugf("Remove state for file as file removed: %s", fe.OldPath)
 
-					err := s.Remove(src.Name())
+					err := s.Remove(src)
 					if err != nil {
 						log.Errorf("Error while removing state from statestore: %v", err)
 					}
 				}
 
 			case loginp.OpRename:
 				log.Debugf("File %s has been renamed to %s", fe.OldPath, fe.NewPath)
-				// TODO update state information in the store
+
+				// if file_identity is based on path, the current reader has to be cancelled
+				// and a new one has to start.
+				if p.identifier.Name() == "path" {
+					prevSrc := p.identifier.GetSource(loginp.FSEvent{NewPath: fe.OldPath})
+					hg.Stop(prevSrc)
+
+					log.Debugf("Remove state for file as file renamed and path file_identity is configured: %s", fe.OldPath)
+					err := s.Remove(prevSrc)
+					if err != nil {
+						log.Errorf("Error while removing old state of renamed file (%s): %v", fe.OldPath, err)
+					}
+
+					hg.Start(ctx, src)
```

Comment:
when calling `Start` here, is it guaranteed that we want to follow this new file? e.g. the ending changed such that the renamed file would not match the grok pattern. In that case the `OpRename` should be handled more siilar to a `OpDelete`.

Reply:
Yes, it is guaranteed. The `prospector` only gets events about files which match the pattern set in `paths`. If a file is renamed to something that is not covered by the pattern, `OpDelete` is returned.